bftengine/tests/controllerWithSimpleHistory/ControllerWithSimpleHistory_test.cpp

```diff
@@ -0,0 +1,133 @@
+
+// Tests methods - insideActiveWindow, onBecomePrimary.
+// Tests logic, which sets the range of sequence numbers that are valid for current window.
+// The starting index is initialized to the next sequence number:
+// that is a multiplication of EvaluationPeriod plus one.
+// The end index is `start + EvaluationPeriod` excluded.
+TEST(ControllerWithSimpleHistory, sequence_number_range_window) {
+  // Test case - init seq is 0.
+  // range should be (1,1+ontrollerWithSimpleHistory::EvaluationPeriod]
+  {
+    // Replica, id = 0 , is primary, view 0, 0 seq num.
+    ControllerWithSimpleHistory testWindow{0, 1, 0, 0, 0};
+    // Should start from one
+    ASSERT_EQ(false, testWindow.insideActiveWindow((SeqNum)0));
+    // End at 1 + EvaluationPeriod, Excluded.
+    ASSERT_EQ(true, testWindow.insideActiveWindow((SeqNum)(1)));
+    // End at 1 + EvaluationPeriod, Excluded.
+    ASSERT_EQ(true, testWindow.insideActiveWindow((SeqNum)(ControllerWithSimpleHistory::EvaluationPeriod / 2)));
+    ASSERT_EQ(true, testWindow.insideActiveWindow((SeqNum)(ControllerWithSimpleHistory::EvaluationPeriod)));
+    // End at 1 + EvaluationPeriod, Excluded.
+    ASSERT_EQ(false, testWindow.insideActiveWindow((SeqNum)(1 + ControllerWithSimpleHistory::EvaluationPeriod)));
+  }
+
+  // Test case - init seq is 122.
+  // range should be next multiplication of EvaluationPeriod plus one:
+  // i.e. (129,129+ontrollerWithSimpleHistory::EvaluationPeriod]
+  {
+    // Replica, id = 0 , is primary, view 0, 122 seq num.
+    ControllerWithSimpleHistory testWindow{0, 1, 0, 0, 122};
+    // Should start from 129
+    ASSERT_EQ(false, testWindow.insideActiveWindow((SeqNum)122));
+    // Should start from 129
+    ASSERT_EQ(true, testWindow.insideActiveWindow((SeqNum)(129)));
+    // End at 129 + EvaluationPeriod, Excluded.
+    ASSERT_EQ(true, testWindow.insideActiveWindow((SeqNum)((129 - 1) + ControllerWithSimpleHistory::EvaluationPeriod)));
+    ASSERT_EQ(false, testWindow.insideActiveWindow((SeqNum)((129) + ControllerWithSimpleHistory::EvaluationPeriod)));
+  }
+}
+
+// Tests method - onNewView.
+// Tests when replica == primary, side effects:
+// 1)Resets slow path timer to default.
+// 2)initializes a range of sequence numbers for the current window to handle.
+// Logic:
+// Run one cycle of EvaluationPeriod requests, which sets the slow path timer to a lower value:
+// then perform onNewView to a view where replica is primary.
+TEST(ControllerWithSimpleHistory, onNewView_is_primary) {
+  ControllerWithSimpleHistory cwsh{0, 1, 0, 0, 0};
+  // Timer value is default
+  auto slowThresh = cwsh.timeToStartSlowPathMilli();
+
+  // Fill one EvaluationPeriod of requests:
+  // Evaluates an average request duration, which is lower than the default.
+  for (auto i = (size_t)1; i <= ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
This is non-deterministic and can cause test cases to fail based on timing. It would be better if instead, you created a  a new function for `cwsh.onNewSeqNumberExecution` that also took a timestamp or duration. You would then call this with wrapper with the current time from a wrapper which is ` cwsh.onNewSeqNumberExecution((SeqNum)i);`

Note that this strategy may require changing some other methods so you can get the current time used, but it shouldn't be much work, and the tests will always pass this way. They also will run faster. In general, anytime you use a sleep in a test, try to see how you can change the code so that the sleep is not needed.

Reply:
I really appreciate this comment.
I tried to test the code with the given interface, I understand better now the expectations.

Reply:
`onNewSeqNumberExecution` measures the duration from the PrePrepare message, I think it would be simpler to implement your suggested strategy to `ControllerWithSimpleHistory::onSendingPrePrepare(SeqNum n, CommitPath commitPath)`  , which all it does is to mark the time point its being called at.
It would also make future tests, planned for that class, simpler and more readable.

Reply:
@andrewjstone  at  `ControllerWithSimpleHistory::onNewSeqNumberExecution` , when When `MIN_DURATION_MICRO  <= duration <= MAX_DURATION_MICRO `
`s.durationMicro_`  is not set i.e. its value is 0.
Is it a bug, it surely affects the avg ?